File: src/Entity/OffreEmploi.php
<?php

namespace App\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;
use App\Entity\Utilisateur;
use App\Entity\Land; // Import Land entity


#[ORM\Entity]
class OffreEmploi
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: "integer")]
    private ?int $id = null; // Primary Key

    #[ORM\Column(type: "string", length: 100)]
    #[Assert\NotBlank(message: "Le titre est obligatoire.")]
    #[Assert\Length(
        min: 3,
        max: 100,
        minMessage: "Le titre doit contenir au moins {{ limit }} caractères.",
        maxMessage: "Le titre ne peut pas dépasser {{ limit }} caractères."
    )]
    private string $titre;

    #[ORM\Column(type: "text")]
    #[Assert\NotBlank(message: "La description est obligatoire.")]
    #[Assert\Length(
        min: 10,
        minMessage: "La description doit contenir au moins {{ limit }} caractères."
    )]
    private string $description;

    #[ORM\Column(type: "float")]
    #[Assert\NotBlank(message: "Le salaire est obligatoire.")]
    #[Assert\Positive(message: "Le salaire doit être un nombre positif.")]
    #[Assert\Range(
        min: 10,
        max: 10000,
        notInRangeMessage: "Le salaire doit être entre {{ min }} et {{ max }} DT."
    )]
    private float $salaire;
    
    #[ORM\Column(type: "string", length: 255)]
    #[Assert\NotBlank(message: "Le lieu est obligatoire.")]
    #[Assert\Length(
        max: 255,
        maxMessage: "Le lieu ne peut pas dépasser {{ limit }} caractères."
    )]
    private string $lieu;

    #[ORM\Column(type: "date")]
    #[Assert\NotBlank(message: "La date de début est obligatoire.")]
    #[Assert\GreaterThan("today", message: "La date de début doit être dans le futur.")]
    private ?\DateTime $startDate = null;

    #[ORM\Column(type: "date")]
    #[Assert\NotBlank(message: "La date d'expiration est obligatoire.")]
    #[Assert\GreaterThan(propertyPath: "startDate", message: "La date d'expiration doit être après la date de début.")]
    private ?\DateTime $dateExpiration = null;

    // ✅ Foreign Key: Link to Utilisateur (Employer/Farmer)
    #[ORM\ManyToOne(targetEntity: Utilisateur::class)]
    #[ORM\JoinColumn(nullable: false, onDelete: "CASCADE")]
    private ?Utilisateur $id_employeur = null;

    /**
     * @var Collection<int, Candidature>
     */
    #[ORM\OneToMany(targetEntity: Candidature::class, mappedBy: 'idOffre', orphanRemoval: true)]
    private Collection $candidatures;


    public function __construct()
    {
        $this->candidatures = new ArrayCollection();
        $this->startDate = new \DateTime(); // Default to today
        $this->dateExpiration = (new \DateTime())->modify('+1 month'); // Default to 1 month from today
    }

    // Getters and Setters
    public function getId(): ?int { return $this->id; }

    public function getTitre(): string { return $this->titre; }
    public function setTitre(string $titre): self { $this->titre = $titre; return $this; }

    public function getDescription(): string { return $this->description; }
    public function setDescription(string $description): self { $this->description = $description; return $this; }

    public function getSalaire(): float { return $this->salaire; }
    public function setSalaire(float $salaire): self { $this->salaire = $salaire; return $this; }

    public function getLieu(): string { return $this->lieu; }
    public function setLieu(string $lieu): self { $this->lieu = $lieu; return $this; }

    public function getStartDate(): ?\DateTime { return $this->startDate; }
    public function setStartDate(?\DateTime $startDate): self { 
        $this->startDate = $startDate ?? new \DateTime(); // Default to today if null
        return $this; 
    }

    public function getDateExpiration(): ?\DateTime { return $this->dateExpiration; }
    public function setDateExpiration(?\DateTime $dateExpiration): self { 
        $this->dateExpiration = $dateExpiration ?? (new \DateTime())->modify('+1 month'); // Default to 1 month ahead if null
        return $this; 
    }

    public function getIdEmployeur(): ?Utilisateur { return $this->id_employeur; }
    public function setIdEmployeur(?Utilisateur $id_employeur): self { $this->id_employeur = $id_employeur; return $this; }

    /**
     * @return Collection<int, Candidature>
     */
    public function getCandidatures(): Collection
    {
        return $this->candidatures;
    }

    public function addCandidature(Candidature $candidature): static
    {
        if (!$this->candidatures->contains($candidature)) {
            $this->candidatures->add($candidature);
            $candidature->setIdOffre($this);
        }

        return $this;
    }

    public function removeCandidature(Candidature $candidature): static
    {
        if ($this->candidatures->removeElement($candidature)) {
            // set the owning side to null (unless already changed)
            if ($candidature->getIdOffre() === $this) {
                $candidature->setIdOffre(null);
            }
        }

        return $this;
    }
    
    public function __toString(): string
    {
        return $this->titre; // Return the job title or any meaningful string representation
    }
}

File: src/main/java/tn/esprit/entities/OffreEmploi.java
package tn.esprit.entities;


import jakarta.persistence.*;
import javax.validation.constraints.*;   // ✅ This is what YOU should use (based on your dependencies)
import java.time.LocalDate;
import java.util.ArrayList;
import java.util.List;

@Entity
public class OffreEmploi {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @Column(length = 100, nullable = false)
    @NotBlank(message = "Le titre est obligatoire.")
    @Size(min = 3, max = 100, message = "Le titre doit contenir entre {min} et {max} caractères.")
    private String titre;

    @Lob
    @NotBlank(message = "La description est obligatoire.")
    @Size(min = 10, message = "La description doit contenir au moins {min} caractères.")
    private String description;

    @Column(nullable = false)
    @NotNull(message = "Le salaire est obligatoire.")
    @Positive(message = "Le salaire doit être un nombre positif.")
    @DecimalMin(value = "10.0", message = "Le salaire doit être au moins {value} DT.")
    @DecimalMax(value = "10000.0", message = "Le salaire doit être au plus {value} DT.")
    private Float salaire;

    @Column(length = 255, nullable = false)
    @NotBlank(message = "Le lieu est obligatoire.")
    @Size(max = 255, message = "Le lieu ne peut pas dépasser {max} caractères.")
    private String lieu;

    @Column(nullable = false)
    @NotNull(message = "La date de début est obligatoire.")
    private LocalDate startDate;

    @Column(nullable = false)
    @NotNull(message = "La date d'expiration est obligatoire.")
    private LocalDate dateExpiration;

    // ManyToOne: Link to User
    @ManyToOne(optional = false, cascade = CascadeType.ALL)
    @JoinColumn(name = "id_employeur_id", nullable = false)
    private User idEmployeur;

    // OneToMany: Candidatures
    @OneToMany(mappedBy = "idOffre", cascade = CascadeType.ALL, orphanRemoval = true)
    private List<Candidature> candidatures = new ArrayList<>();

    public OffreEmploi() {
        this.startDate = LocalDate.now();
        this.dateExpiration = LocalDate.now().plusMonths(1);
    }

    // Getters and Setters
    public Integer getId() { return id; }

    public String getTitre() { return titre; }
    public void setTitre(String titre) { this.titre = titre; }

    public String getDescription() { return description; }
    public void setDescription(String description) { this.description = description; }

    public Float getSalaire() { return salaire; }
    public void setSalaire(Float salaire) { this.salaire = salaire; }

    public String getLieu() { return lieu; }
    public void setLieu(String lieu) { this.lieu = lieu; }

    public LocalDate getStartDate() { return startDate; }
    public void setStartDate(LocalDate startDate) {
        this.startDate = (startDate != null) ? startDate : LocalDate.now();
    }

    public LocalDate getDateExpiration() { return dateExpiration; }
    public void setDateExpiration(LocalDate dateExpiration) {
        this.dateExpiration = (dateExpiration != null) ? dateExpiration : LocalDate.now().plusMonths(1);
    }

    public User getIdEmployeur() { return idEmployeur; }
    public void setIdEmployeur(User idEmployeur) { this.idEmployeur = idEmployeur; }

    public List<Candidature> getCandidatures() { return candidatures; }

    public void addCandidature(Candidature candidature) {
        if (!this.candidatures.contains(candidature)) {
            this.candidatures.add(candidature);
            candidature.setIdOffre(this);
        }
    }

    public void removeCandidature(Candidature candidature) {
        if (this.candidatures.remove(candidature)) {
            if (candidature.getIdOffre() == this) {
                candidature.setIdOffre(null);
            }
        }
    }

    @Override
    public String toString() {
        return this.titre;
    }
    public void setId(int id) {
        this.id = id;
    }

}
